optuna/visualization/matplotlib/_optimization_history.py

```diff
@@ -128,11 +130,98 @@ def _get_optimization_history_plot(
         _logger.warning("Study instance does not contain trials.")
         return ax
 
-    ax = _get_optimization_histories(studies, target, target_name, ax)
+    if error_bar:
+        ax = _get_optimization_histories_with_error_bar(studies, target, target_name, ax)
+    else:
+        ax = _get_optimization_histories(studies, target, target_name, ax)
+
     plt.legend(bbox_to_anchor=(1.05, 1.0), loc="upper left")
     return ax
 
 
+def _get_optimization_histories_with_error_bar(
+    studies: List[Study],
+    target: Optional[Callable[[FrozenTrial], float]],
+    target_name: str,
+    ax: "Axes",
+) -> "Axes":
+
+    max_trial_number = np.max(
+        [
+            trial.number
+            for study in studies
+            for trial in study.get_trials(states=(TrialState.COMPLETE,))
+        ]
+    )
+
+    _target: Callable[
+        [
+            FrozenTrial,
+        ],
+        float,
+    ]
+    if target is None:
+
+        def _target(t: FrozenTrial) -> float:
+            return cast(float, t.value)
+
+    else:
+        _target = target
+
+    target_values: List[List[float]] = [[] for _ in range(max_trial_number + 2)]
+    for study in studies:
+        trials = study.get_trials(states=(TrialState.COMPLETE,))
+        for t in trials:
+            target_values[t.number].append(_target(t))
+
+    mean_of_target_values = [np.mean(v) if len(v) > 0 else None for v in target_values]
+    std_of_target_values = [np.std(v) if len(v) > 0 else None for v in target_values]
+    trial_numbers = np.arange(max_trial_number + 2)[[v is not None for v in mean_of_target_values]]
+    means = np.asarray(mean_of_target_values)[trial_numbers]
+    stds = np.asarray(std_of_target_values)[trial_numbers]
+
+    plt.errorbar(
+        x=trial_numbers,
+        y=means,
+        yerr=stds,
+        capsize=5,
+        fmt="o",
+        color="tab:blue",
+    )
+    ax.scatter(trial_numbers, means, color="tab:blue", label=target_name)
+
+    best_values: List[List[float]] = [[] for _ in range(max_trial_number + 2)]
+
+    if target is None:
```

Comment:
Since`best_values` is necessary when `target` is `None`, how about this?

```suggestion
    
    if target is None:
        best_values: List[List[float]] = [[] for _ in range(max_trial_number + 2)]
```